Userland/DevTools/RelationalStudio/ScriptEditor.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022, Dylan Katz <dykatz@uw.edu>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/File.h>
+#include <LibGUI/Dialog.h>
+#include <LibGUI/FilePicker.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/TabWidget.h>
+#include <LibSQL/AST/SyntaxHighlighter.h>
+
+#include "ScriptEditor.h"
+
+namespace RelationalStudio {
+
+ScriptEditor::ScriptEditor()
+{
+    set_syntax_highlighter(make<SQL::AST::SyntaxHighlighter>());
+    set_ruler_visible(true);
+}
+
+void ScriptEditor::new_script_with_temp_name(String name)
+{
+    set_name(name);
+}
+
+void ScriptEditor::open_script_from_file(LexicalPath const& file_path)
+{
+    auto file_or_error = Core::File::open(file_path.string(), Core::OpenMode::ReadOnly);
+    if (file_or_error.is_error()) {
+        auto error = file_or_error.release_error();
+        GUI::MessageBox::show(window(), String::formatted("Couldn't open {}\n{}", file_path, error), "Error", GUI::MessageBox::Type::Error);
+        return;
+    }
+    auto file = file_or_error.release_value();
+    set_text(file->read_all());
+
+    m_path = file_path.string();
+    set_name(file_path.title());
+}
+
+bool ScriptEditor::save()
```

Comment:
Wouldn't the error reporting on these be better with `ErrorOr<void>`? You could just have a single `GUI::MessageBox::show` call with the error's string. Also it looks like save() always fallbacks to save_as, so you could handle that centrally as well.

Reply:
I can investigate using `ErrorOr<void>` in these places, but these aren't necessarily errors. If the user dismisses a save, that isn't an error, and if there is an actual error opening/saving a file, that is already handled with a message box by the time the function exits. The return value is just to do with the program flow.

Though, I may want to try hoisting the error dialog up into the main widget and out of the `ScriptEditor` class, in which case the return value should be `ErrorOr<bool>`.